<h1><b>Waze User Churn Project </b></h1>

<h3>Capstone Project: Google Advanced Data Analytics Professional Certificate</h3>

<h2><b> Background and Business Problem</b></h2>
<p>Waze's navigation app serves millions of drivers worldwide, providing free navigation services through a collaborative community of users, map editors, and partners. However, like many digital platforms, Waze faces the challenge of user churn—users who uninstall the app or stop using it entirely. This creates significant business challenges that directly impact growth and sustainability.

The manual analysis of user behavior patterns was time-consuming and reactive, leading to several critical issues:

- <b>Reactive churn management:</b> Users were already lost before intervention strategies could be implemented
- <b>Inefficient resource allocation:</b> Marketing and retention efforts were spread broadly rather than targeted at high-risk users
- <b>Limited predictive capability:</b> Inability to identify early warning signs of potential churn
- <b>Missed retention opportunities: </b> High-value users at risk of churning went unidentified until it was too late

To address these challenges and support Waze's growth objectives, the data team aims to develop a predictive machine learning model that can identify users at high risk of churning before they actually leave the platform. This proactive approach will enable targeted retention strategies and improve overall user satisfaction.</p>

<h2><b>Business Impact</b></h2>
<p>Implementing a predictive model to identify users at high risk of monthly churn will have direct and measurable impact on Waze's business performance and user retention strategy.</p>
<ul>
    <li><b>Proactie user retention:</b> Early identiifcaton of at-risk users enables targeted intervention before churn occurs. </li>
    <li><b>Optimized vmarketing spend: </b> Resources can be focused on high -risk, high-vvalue users rather than broad based campaigns </li>
    <li><b>Improved user experience: </b>Understanding churn factors helps product teams address pain points that drive users away. </li>
    <li><b>Enhaced revenue potential: </b>Higher retention rates directly correlate with increased user lifetime value and advertising revenu. </li>
    <li><b>Data-driven decision making: </b>Quantifiable insights replace intuition based retetion strategies.</li>
</ul>




<h2><b>Methodology</b></h2>
<h4>1) Data Collection and Preparation </h4>
<ul>
    <li>The goal is to construct a dataframe in Python, perform initial inspection of the provided dataset, and inform  data team members of the findings.</li>
    <li> Understanding the situation: How can the user behaviour data be prepared and organized to identify churn patterns. </li>
    <li> Understand the data: Creating necessary dataframes to understand the user engagement metrics and patterns for future EDA and statistical analysis. </li>
    <li>Understanding variables:  Using insights from the examination of the user activity data to guide deeper investigation into churn prediction variables</li>
    
</ul>

In [6]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

In [7]:
df = pd.read_csv("waze_dataset.csv")
df.head(10)

,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android
5,5,retained,113,103,279.544437,2637,0,0,901.238699,439.101397,15,11,iPhone
6,6,retained,3,2,236.725314,360,185,18,5249.172828,726.577205,28,23,iPhone
7,7,retained,39,35,176.072845,2999,0,0,7892.052468,2466.981741,22,20,iPhone
8,8,retained,57,46,183.532018,424,0,26,2651.709764,1594.342984,25,20,Android
9,9,churned,84,68,244.802115,2997,72,0,6043.460295,2341.838528,7,3,iPhone


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [9]:
df.isna().sum()

ID                           0
label                      700
sessions                     0
drives                       0
total_sessions               0
n_days_after_onboarding      0
total_navigations_fav1       0
total_navigations_fav2       0
driven_km_drives             0
duration_minutes_drives      0
activity_days                0
driving_days                 0
device                       0
dtype: int64

The following can be observed from above:
1) From the first 10 rows we can see that two categorical variables while all other columns are numerical in nature.
   
2) There are total of 14,999 rows and 13 columns. Out of that there are 700 missing values in label column while columns does not have any missing values.

 
3) The `label` and `device` columns are of type `object` while `total_sessions`, `driven_km_drives` and `duration_minutes_drives` columns are of type `float64` while all other columns are of type `int64`.

<b>Null values and summary statistics</b>


In [10]:
# summary statistics of null values in label columns
null_label_df = df[df.label.isnull()]
null_label_df.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000
mean,7405.584286,80.837143,67.798571,198.483348,1709.295714,118.717143,30.371429,3935.967029,1795.123358,15.382857,12.125714
std,4306.900234,79.987440,65.271926,140.561715,1005.306562,156.308140,46.306984,2443.107121,1419.242246,8.772714,7.626373
min,77.000000,0.000000,0.000000,5.582648,16.000000,0.000000,0.000000,290.119811,66.588493,0.000000,0.000000
25%,3744.500000,23.000000,20.000000,94.056340,869.000000,4.000000,0.000000,2119.344818,779.009271,8.000000,6.000000
50%,7443.000000,56.000000,47.500000,177.255925,1650.500000,62.500000,10.000000,3421.156721,1414.966279,15.000000,12.000000
75%,11007.000000,112.250000,94.000000,266.058022,2508.750000,169.250000,43.000000,5166.097373,2443.955404,23.000000,18.000000
max,14993.000000,556.000000,445.000000,1076.879741,3498.000000,1096.000000,352.000000,15135.391280,9746.253023,31.000000,30.000000


In [11]:
# summary statistics for non null values in label columns
non_null_label_df = df[~df.label.isnull()]
non_null_label_df.describe()

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000
mean,7503.573117,80.623820,67.255822,189.547409,1751.822505,121.747395,29.638296,4044.401535,1864.199794,15.544653,12.182530
std,4331.207621,80.736502,65.947295,136.189764,1008.663834,147.713428,45.350890,2504.977970,1448.005047,9.016088,7.833835
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3749.500000,23.000000,20.000000,90.457733,878.500000,10.000000,0.000000,2217.319909,840.181344,8.000000,5.000000
50%,7504.000000,56.000000,48.000000,158.718571,1749.000000,71.000000,9.000000,3496.545617,1479.394387,16.000000,12.000000
75%,11257.500000,111.000000,93.000000,253.540450,2627.500000,178.000000,43.000000,5299.972162,2466.928876,23.000000,19.000000
max,14998.000000,743.000000,596.000000,1216.154633,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


Observations from above: 
1) There is a large differnce in the mean and median values amongst all columns which indicates the presence of outliers and hence mean is not a good indicator of the overall characteristic of the data.
   
2) There is no difference between the summary statistics of the missing retention values and those that are not missing. The mean and standard deviation are fairly consistent between the two groups.

3) The maximum value of `driven_km_drives` is more than half the circumference of the earth! has to be a serious driver! need to further look into this.

4) Maximum driving days is 30 days which indicate that this is a monthly data which is correct. 

In [23]:
df.head(5)

,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device,km_per_drive,km_per_driving_day,drives_per_driving_day
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android,11.632058,138.360267,11.894737
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone,128.186173,1246.901868,9.727273
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android,32.201567,382.393602,11.875000
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone,22.839778,304.530374,13.333333
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android,58.091206,219.455667,3.777778


<b>From null label values - device counts</b>

In [12]:
null_label_df.device.value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

Out of the rows with missing values (700), there are 447 iPhone users and 253 Android users. 

In [24]:
# what percentage does users of each category of device contribute ? 

null_label_df.device.value_counts()/700

device
iPhone     0.638571
Android    0.361429
Name: count, dtype: float64

From the missing rows (700), 63% users use iPhone and 36% use Android device.

In [25]:
# now to see what the ratio is in full dataset
df.device.value_counts()/len(df.device)

device
iPhone     0.644843
Android    0.355157
Name: count, dtype: float64

The percentage of missing value by each device is consistent with the device count of overall data. 

In [28]:
# What is the number of users who churned vs retained out of full dataset ? 
print(df.label.value_counts())
print()
print("Percentage \n")
print(df.label.value_counts(normalize=True))

label
retained    11763
churned      2536
Name: count, dtype: int64

Percentage 

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64


This dataset contains 82% retained users and 18% churned users. 

To understand a bit more about the each variables for churned and retained users we will calculate the median as mean is affected by outliers. 

In [16]:
df.groupby(['label']).median(numeric_only=True)

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
label,,,,,,,,,,,
churned,7477.5,59.0,50.0,164.339042,1321.0,84.5,11.0,3652.655666,1607.183785,8.0,6.0
retained,7509.0,56.0,47.0,157.586756,1843.0,68.0,9.0,3464.684614,1458.046141,17.0,14.0


The above gives a good snapshot of two groups, churned vs. retained: 

1) The users who churned had averaged 3 more drives ( from `drives` ) than the retained users but retained users use the app twice as many times as churned users (from `activity_days` ) in the same time period.
   
2) The median churned users drove ~200 more kilometres (from `driven_km_drives` ) and 2.5 hours more than the median retained user (from `duration_minutes_drives`).

   
3) Overall, it looks like churn users have more drives in fewer days and have driven more and for longer hours than the retained users. We can look more into this by taking a median kilometers per drive for both retained and churned users. 
   

In [32]:
# How much average kilometres were driven per drive ?
df['km_per_drive'] = df.driven_km_drives / df.drives 
median_km_per_drive = df.groupby('label').agg({'km_per_drive': 'median'})
median_km_per_drive

,km_per_drive
label,
churned,74.109416
retained,75.014702


The median retained user drove about one kilometer more on average per drive than the median churned user. 

In [33]:
# How much average kilometers were driven per driving day ?   
df['km_per_driving_day'] = df.driven_km_drives / df.driving_days
median_km_per_driving_day = df.groupby('label').median(numeric_only=True)[['km_per_driving_day']]
median_km_per_driving_day

,km_per_driving_day
label,
churned,697.541999
retained,289.549333


To understand the median number of drives per driving day for each group

In [20]:
df['drives_per_driving_day'] = df.drives / df.driving_days
median_drives_per_driving_day = df.groupby('label').median(numeric_only=True)[['drives_per_driving_day']]
median_drives_per_driving_day

,drives_per_driving_day
label,
churned,10.0000
retained,4.0625


From above we can observe that: 

1) The median users who churned drove 698 kilometres each day than they drove last month which is **~240%** higher than the retained users. This is a lot of driving per drive day.
   
2) The median churned user had similarly disproportionate numbers of drives per drive day compared to the retained users, almost more than twice of retained users. 

Regardless of churned users or not the data represent users who are serious drivers! it would be safe to assume that this data does not represent typical drivers at large. Perhaps the data and particularly churned users data contains a high proportion of long-haul truckers. 

This information is worth mentioning to Waze as more data needs to be gathered on these serious drivers. It is possible that the reason for them driving so much is also the reason why Waze app does not meet specific requirement or set of needs, which differes from the need of a typical driver such as a commuter. This might also suggest a non random sampling where the sample does not represent typical drivers and only focus on long haul drivers. 


In [21]:
# understanding the count of devices of churned vs retained users 
df.groupby(['label', 'device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [22]:
# count of device by user type but in percentage
df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

The ratio of iPhone users and androd users is consistent between churned and retained groups and the ratio are both consistent with the ratio found in the overall dataset. 


<h4>2) Exploratory Data Analysis </h4>
<ul>
    <li>The purpose of this part is to conduct exploratory data analysis on the dataset. </li>
    <li> The goal is to examine the data that we began with previously, adding relevant visualization that can help communicate the story that the data tells.  </li>
    <li><b>Understand the data: </b>Perform EDA on important columns. Find outliers and handle outliers. Some necessary visualizations to show to stakeholders and creation of dashboard to help non technical stakeholders undestand data with accessibility in mind. </li>
 <li><b>Understanding variables: </b>Leverage insights from previous summary statistics and data examination to guide deeper investigation into key retention-determining variables</li>
    
</ul>